### PR TITLE
Avoid NPE when selecting hints and consistency levels

### PR DESCRIPTION
**Pull request by apmoriarty**
- avoid NPE when selecting hints and consistency levels
- expanded ScannerFactory test to account for non-configured tables
- expanded ScannerFactory trace logging to include invalid hints

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java [hunk 1]
```diff
@@ -135,7 +136,7 @@ public void updateConfigs(GenericQueryConfiguration genericConfig) {
         }
 
         if (log.isDebugEnabled()) {
-            log.debug("Created ScannerFactory " + System.identityHashCode(this) + " is wrapped ? " + (client instanceof WrappedConnector));
+            log.debug("Created ScannerFactory {} is wrapped? {}", System.identityHashCode(this), (client instanceof WrappedConnector));
```
Comment (avgAGB) [inline warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]: Not your fault, but the original grammar for this message seems a little confusing. Maybe it'd be worth separating the clauses into something like "Created ScannerFactory {}. Is the factory wrapped: {}"

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java [hunk 2]
```diff
@@ -358,7 +359,7 @@ public <T extends ScannerSession> T newLimitedScanner(Class<T> wrapper, final St
         Preconditions.checkNotNull(wrapper);
         Preconditions.checkArgument(open.get(), "Factory has been locked. No New scanners can be created");
 
-        log.debug("Creating limited scanner whose max threads is is " + scanQueue.getCapacity() + " and max capacity is " + maxQueue);
+        log.debug("Creating limited scanner whose max threads is is {} and max capacity is {}", scanQueue.getCapacity(), maxQueue);
```
Comment (avgAGB) [inline warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]: extra "is" in the message here.

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java [hunk 3]
```diff
@@ -539,13 +540,29 @@ public void applyConfigs(ScannerBase scannerBase, String tableName) {
...
+            } else {
+                scannerBase.setConsistencyLevel(level);
+            }
         }
 
         if (executionHintMap != null && !executionHintMap.isEmpty()) {
-            String key = executionHintMap.containsKey(hintKey) ? hintKey : tableName;
-            scannerBase.setExecutionHints(executionHintMap.get(key));
+            Map<String,String> hint = executionHintMap.get(hintKey);
+            if (hint == null) {
```
Comment (avgAGB) [inline warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java:557]: Is it possible to combine the log trace statement in this first clause? Same question would apply for the other if statements below.

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]: Updated that line to be more clear

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java (hunk 2 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java]: Fixed duplicate

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java (hunk 3 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java:557]: I think those log statements should stay separate for clarity. It's quite possible to have consistently levels configured correctly, but the scan hint is incorrect and vice versa.

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java [hunk 4]
```diff
@@ -563,18 +580,28 @@ protected void applyConfigs(ScannerSession scannerSession, String hintKey, Strin
...
+                hint = executionHintMap.get(tableName);
+            }
 
-            if (executionHintMap.containsKey(key)) {
-                options.setExecutionHints(executionHintMap.get(key));
+            if (hint == null) {
+                log.trace("no execution hint found for table: {} key: {} ", tableName, hintKey);
+            } else {
+                options.setExecutionHints(hint);
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java:604]: this looks like the same code as the block around ~565, please use a method.

**Review by alerman — APPROVED** [review]

**Review by keith-ratcliffe — APPROVED** [review]

File: warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java (hunk 4 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java:604]: good call -- commit forthcoming

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/tables/ScannerFactory.java:604]: Commit can be found  here: https://github.com/NationalSecurityAgency/datawave/pull/2639

**Pull request merged**